lib/mlir/Dialect/Clift/IR/CliftOps.cpp

```diff
@@ -273,15 +312,48 @@ struct ModuleValidator {
                                   "ModuleOp.";
     }
 
+    if (auto F = mlir::dyn_cast<FunctionOp>(Op)) {
+      auto TypeAttr = getFunctionTypeAttr(F.getFunctionType());
+      FunctionReturnType = mlir::cast<ValueType>(TypeAttr.getReturnType());
+    }
+
     return visitOp(Op);
   }
 
 private:
   clift::ModuleOp Module;
   Operation *ModuleLevelOp;
+  clift::ValueType FunctionReturnType;
   llvm::SmallPtrSet<mlir::Type, 32> VisitedTypes;
   llvm::SmallPtrSet<mlir::Attribute, 32> VisitedAttrs;
   llvm::DenseMap<uint64_t, TypeDefinitionAttr> Definitions;
+
+  static std::optional<ControlStatementKind> getStatementKind(Operation *Op) {
+    if (mlir::isa<ForOp, DoWhileOp, WhileOp>(Op))
+      return ControlStatementKind::Loop;
+
+    if (mlir::isa<SwitchOp>(Op))
+      return ControlStatementKind::Switch;
+
+    return std::nullopt;
+  }
+
+  bool hasParentStatement(Operation *Op,
+                          ControlStatementKind Kind,
+                          bool DirectlyNested = true) {
```

Comment:
I defaulted it to `true` because a) that's the more conservative option, i.e. more likely to not find the required parent, and b) it's used more often.

Still, I've removed the default argument.

Reply:
Thanks.